### DISCUSSION TITLE
Drop cache for dynamically generated queries.

### DISCUSSION THREAD

**Discussion by Geddard** [urql-graphql/urql#1336] [Q&A] [answered]
Hi there, I've been reading up on the docs and trying to find other people with my issues, but I can't figure out how to do this... I either end up with no caching at all or just not able to invalidate it.

The scenario is that I call my graphql endpoint with dynamically generated queries, i never know what are the contents of these, at mas i know which is a mutation or which is a normal query. But i can't know the names of the internal values.

From what i see in the docs, i can only invalidate cache with these exchanges (types don't work if I try to use the graphcache btw) but i need to specify the query key i want to drop the cache for... in my case i would like to just drop ALL cache one i do a mutation.

I also tried calling `createClient({ url })` again after a mutation, but i get errors when i try to do another query after that, about some numeric id, or i just don't see any changes.

I have a recoil atom where i store the client

```
export const dataClientState = atom({
  key: 'dataClient',
  default: createClient({ url: DATA }),
});
```

So then i created a hook where i reset this state by recreating the client, but the cache doesn't seem to go away:

```
export const useMutation = () => {
  const [dataClient, resetDataClient] = useRecoilState(dataClientState);

  const clientRef = useRef(dataClient);

  useEffect(() => {
    clientRef.current = dataClient;
  }, [clientRef, dataClient, resetDataClient]);

  return <T = unknown>(query: DocumentNode) =>
    from(dataClient.mutation<T>(query).toPromise()).pipe(
      map(response => {
        resetDataClient(createClient({ url: DATA }));

        return response;
      })
    );
};
```

I gues that the ideal would be to be able to pass part of the dynamically generated query to mutation() and that should invalidate part of the cache? if not all??

I'm lost here and considering writing URQL out of my project... please help

**Comment by Geddard** [comment]
UPDATE: I thought it might an issue of the closure returned in the hook, but i extracted all the function and just pasted the call where i need it and the problem remains, calling createClient doesn't seem to invalidate cache...

**Comment by JoviDeCroock** [comment]
All right let's look at the problem first, what triggers this data to be different every time?

If it's you interacting with the data then a useMutation is needed to show these interactions.

If it's other people interacting with the data we will need a useSubscription.

If this one query is just truely dynamic try using fetchPolicy: 'network-only'.

What makes the query dynamic? Resetting the client shouldn't be needed? If possible reproduce it on codesandbox and we'll gladly point you in the right direction.

Sidenote: what about the graphCache types isn't working? Just tried it and worked.

**Reply by Geddard** [reply]
> All right let's look at the problem first, what triggers this data to be different every time?
> 
> If it's you interacting with the data then a useMutation is needed to show these interactions.
> 
> If it's other people interacting with the data we will need a useSubscription.
> 
> If this one query is just truely dynamic try using fetchPolicy: 'network-only'.
> 
> What makes the query dynamic? Resetting the client shouldn't be needed? If possible reproduce it on codesandbox.
> 
> Sidenote: what about the graphCache types isn't working? Just tried it and worked.

Ok, let's ignore the types issue for now, maybe i forgot something there.

But i'll try to explain my situation.

Im developing an SPA, so let's say I'm right now on a page that shows details for a product.

In this page i can edit fields of the information displayed, for example the "product code", but this field is a special autocomplete that shows you a code with an incremented number, based on what is already in the DB.

So for example, these codes are in the DB: `code001, code002, code003`, so when i type "code" in the autocomplete, i do a search for that value and i get "suggested code: code004, based on code003". (gql query performed)

Then i select this value, and click save (mutation). After this, the page doesn't refresh, i just update the data with the mutation response.
But when i try to search in the code field again for "code", i still have the cached response that will show me "use code 004".



The queries i perform are dynamic because this SPA has many different "Data sections" that have similar queries but with different names, but the config for the queries is in the backend, frontend only grabs the config, renders and performs actions. So there's no way i could configure all the possible queries i need to invalidate cache for at the start of the app.

I'm not sure how it works, but in apollo it seems i can send an extra param to mutations that can invalidate cache (https://github.com/lucasconstantino/apollo-cache-invalidation#readme) is there something similar?

At this point i wouldn't mind some "hacky" solution for invalidating cache

Thanks for the help!

**Reply by JoviDeCroock** [reply]
If you are using the document-cache you van use `additionalTypenames`, however I think applying a fetchPolicy of `network-only` to the search query should be sufficient.

**Reply by Geddard** [reply]
@JoviDeCroock I considered that, but i need cache sometimes, for example when searching for thousands of results... i thought i could somehow configure many queries to invalidate related to the mutations, but it's too much hassle for little benefit. 
I wish there was just a way of dropping all cache at once after a mutation

**Reply by JoviDeCroock** [reply]
Yes, `additionalTypenames` is exactly that. You can specify typenames that relate to what needs to be invalidated https://formidable.com/open-source/urql/docs/basics/document-caching/#adding-typenames. If you want updater methods like Apollo, graphcache is what you need.

**Reply by Geddard** [reply]
> Yes, `additionalTypenames` is exactly that. You can specify typenames that relate to what needs to be invalidated https://formidable.com/open-source/urql/docs/basics/document-caching/#adding-typenames. If you want updater methods like Apollo, graphcache is what you need.

But then this means if I need to invalidate 1000 different possible queries then i have to list them in those typenames?
As i said, im not sure how Apollo works, i haven't tried it yet. But are you saying that with graphcache i can drop 100% of the cache after mutations?

**Reply by JoviDeCroock** [reply]
No, I think this is a pretty edge-casey situation. Resetting the client seems to work here, https://codesandbox.io/s/clever-pine-tcpq7?file=/src/index.js how can I get to the error you were seeing? 

You could look at providing an additional context here that exposes the `resetClient` to the whole React tree.

**Reply by Geddard** [reply]
Man this is complicated... i just tried manually adding the typenames to test if it would work and no luck, if the query looks the same, the results are the same... my only idea at this point is to add some kind of cache buster that i can append to the query

**Reply by Geddard** [reply]
@JoviDeCroock Btw this is one of the errors i get when i try to use the graphcache:

here 

```
import { cacheExchange } from '@urql/exchange-graphcache';
import { atom } from 'recoil';
import { createClient, dedupExchange, fetchExchange } from 'urql';

export const DATA = '/data';

export const dataClientState = atom({
  key: 'dataClient',
  default: createClient({ url: DATA, exchanges: [dedupExchange, cacheExchange({}), fetchExchange] }),
});

```

cacheExchange is underlined in red and i get an error wth this text:

> Type 'import("/.../frontend/node_modules/@urql/core/dist/types/types").Exchange' is not assignable to type 'import("/.../frontend/node_modules/urql/node_modules/@urql/core/dist/types/types").Exchange'.
  Types of parameters 'input' and 'input' are incompatible.
    Type 'import("/.../frontend/node_modules/urql/node_modules/@urql/core/dist/types/types").ExchangeInput' is not assignable to type 'import("/.../frontend/node_modules/@urql/core/dist/types/types").ExchangeInput'.
      Types of property 'client' are incompatible.
        Type 'import("/.../frontend/node_modules/urql/node_modules/@urql/core/dist/types/client").Client' is not assignable to type 'import("/.../frontend/node_modules/@urql/core/dist/types/client").Client'.
          Types have separate declarations of a private property 'onOperationStart'.ts(2322)

**Reply by Geddard** [reply]
I can only get it to "work" if i cast  `as any`

**Reply by Geddard** [reply]
I also just tried this:

```
cacheExchange({
        updates: {
          Mutation: {
            MainQueryName: (result, args, cache, info) => {
              cache.invalidate({ __typename: 'MainQueryName', id: args.id as string }); // i have to cast as string otherwise i get type errors
            },
          },
        },
      }) as any,
```

And when i try to do a mutation i get this error:

>Unhandled Rejection (Graphcache Error): [object Object]`.
(Caused At: Unnamed mutation)
https://bit.ly/2XbVrpR#19

**Reply by Geddard** [reply]
That config should match this query, right?

```
mutation {
 MainQueryName {
    innerFields...
 }
}
```

**Reply by Geddard** [reply]
Since there seems to be no light at the end of the tunnel, i'll be switching off to relay, there's a beautiful `cache.clear()` function there :)
Thanks for the help anyway

**Reply by kitten** [reply]
That error is unrelated. It occurs when you've got duplicate versions of `@urql/core`installed. You can resolve that with `yarn-deduplicate` or `npm dedupe` depending on your package manager.

The problem why this isn't going far is that you're writing what you want to do, get a couple of suggestions and questions back, and then run into a problem and post that; rinse and repeat.

If we knew more about what you're trying to achieve before you list the problems in the details then we can explain where the misunderstandings are potentially. For instance, when do you need fresh data? And when do you fetch? And what's the condition for having thousands of cached queries here. That's unusual; so how are
You expecting to retrieve them?

Also it does sound like you 100% need Graphcache. Without a normalised cache a dynamically generated query will only be cached as a document which wouldn't be very effective. But again, I can only guess what your use-case is exactly

**Reply by Geddard** [reply]
> That error is unrelated. It occurs when you've got duplicate versions of `@urql/core`installed. You can resolve that with `yarn-deduplicate` or `npm dedupe` depending on your package manager.
> 
> The problem why this isn't going far is that you're writing what you want to do, get a couple of suggestions and questions back, and then run into a problem and post that; rinse and repeat.
> 
> If we knew more about what you're trying to achieve before you list the problems in the details then we can explain where the misunderstandings are potentially. For instance, when do you need fresh data? And when do you fetch? And what's the condition for having thousands of cached queries here. That's unusual; so how are
> You expecting to retrieve them?
> 
> Also it does sound like you 100% need Graphcache. Without a normalised cache a dynamically generated query will only be cached as a document which wouldn't be very effective. But again, I can only guess what your use-case is exactly

 I understand, I've tried explaining the problem as much as possible, but it boils down to this:

 Perform any mutation -> DROP ALL CACHE

That's all I need to do, there's no other solution that works for my project because of its nature. Since apparently, urql doesn't have any way of dropping all cache at once, my only option is to switch to another library ;(

**Reply by kitten** [reply]
@Geddard You can imitate this on Graphcache by using an updater that calls `cache.invalidate("Query")` after your mutation, or just drop everything in the document cache with `additionalTypenames` as mentioned.

The reason why I'm a bit suspect though is because you've already arrived at a conclusion without the underlying problem which limits our suggestions.

If you're trying to prevent stale data for instance, there are other mechanisms that can prevent this: https://formidable.com/open-source/urql/docs/basics/queries/#request-policies

Overall though I'd say that invalidating everything without any particular trigger or connection to the data from the mutation to the query is outside the scope of regular use-cases, which is why our suggestions will always be of limited use 😅

**Reply by Geddard** [reply]
> @Geddard You can imitate this on Graphcache by using an updater that calls `cache.invalidate("Query")` after your mutation, or just drop everything in the document cache with `additionalTypenames` as mentioned.
> 
> The reason why I'm a bit suspect though is because you've already arrived at a conclusion without the underlying problem which limits our suggestions.
> 
> If you're trying to prevent stale data for instance, there are other mechanisms that can prevent this: https://formidable.com/open-source/urql/docs/basics/queries/#request-policies
> 
> Overall though I'd say that invalidating everything without any particular trigger or connection to the data from the mutation to the query is outside the scope of regular use-cases, which is why our suggestions will always be of limited use 😅

As i mentioned before, i tried both, i get either errors or the cache stops working entirely (not desired)
Could you maybe check the code i pasted for those and help me figure out how to config this graphcache to invalidate?

>Overall though I'd say that invalidating everything without any particular trigger or connection to the data from the mutation to the query is outside the scope of regular use-cases, which is why our suggestions will always be of limited use 😅

I understand that is not regular, but i also don't see why there isn't an easy `cache.clear() ` method somewhere... I'm checking other libraries and most do have something like this

**Reply by kitten** [reply]
The error above is caused because you've used the name of your mutation. The first thing that stands out is that the mutation is actually unnamed.
The second is that you have to set `updates.Mutation[fieldName]` so that `fieldName` is the mutation field. There's more on that in the docs.

As this is a normalised cache it'll start caching by fields starting from `Query`, so everything that can be reached from a query, and will always deal with field names in general.

There is no "invalidate" all call for several reasons, one of them being that a cache that loses all data at all times follows patterns that don't go well with GraphQL.
People have asked similar questions around authentication and our recommendation then is to actually recreate the entire client in these cases which clears more data.

**Reply by Geddard** [reply]
> The error above is caused because you've used the name of your mutation. The first thing that stands out is that the mutation is actually unnamed.
> The second is that you have to set `updates.Mutation[fieldName]` so that `fieldName` is the mutation field. There's more on that in the docs.
> 
> As this is a normalised cache it'll start caching by fields starting from `Query`, so everything that can be reached from a query, and will always deal with field names in general.
> 
> There is no "invalidate" all call for several reasons, one of them being that a cache that loses all data at all times follows patterns that don't go well with GraphQL.
> People have asked similar questions around authentication and our recommendation then is to actually recreate the entire client in these cases which clears more data.

Yeah i really don't understand still how to solve the graphcache problems, i also don't think it make sense that i "have to" name my queries, so that's another big drawback of this library i guess.

I'm also still trying to recreate the client, i don't get any errors but also don't loose the cache...

```
import { DocumentNode } from 'graphql';
import { useEffect, useRef, useState } from 'react';
import { from } from 'rxjs';
import { map } from 'rxjs/operators';
import { createClient } from 'urql';
export const DATA = '/data';
export const CONFIG = '/graphql';

export const useMutation = () => {
  const [shouldResetState, setShouldResetState] = useState<boolean>();
  const dataClient = useRef(createClient({ url: DATA }));

  useEffect(() => {
    if (shouldResetState) {
      dataClient.current = createClient({ url: DATA });
      setShouldResetState(false);
    }
  }, [shouldResetState]);

  return <T = unknown>(query: DocumentNode) =>
    from(dataClient.current.mutation<T>(query).toPromise()).pipe(
      map(response => {
        setShouldResetState(true);

        return response;
      })
    );
};
```

**Reply by JoviDeCroock** [reply]
Because you're not resetting the Provider, see my [example](https://codesandbox.io/s/clever-pine-tcpq7?file=/src/index.js)

**Reply by Geddard** [reply]
> Because you're not resetting the Provider, see my [example](https://codesandbox.io/s/clever-pine-tcpq7?file=/src/index.js)

@JoviDeCroock  I'm not using providers at all because i have more than one graphql URL i need to access, and so far just creating the client and using it (as you can see in my code) directly with `clien.mutation()` works. 

Anything else i can try?

**Reply by JoviDeCroock** [reply]
I mean, your cache is scoped to your client so if you mutate on clientX you can't expect clientY to invalidate that won't work in any client.... We have `context.url` to use different urls for hooks.

**Reply by Geddard** [reply]
> I mean, your cache is scoped to your client so if you mutate on clientX you can't expect clientY to invalidate that won't work in any client.... We have `context.url` to use different urls for hooks.

I'm only talking about invalidating the cache in the same client, not another one.

Also, i just tried this config and the cache simply doesn't work:

```

export const dataClientState = atom({
  key: 'dataClient',
  default: createClient({
    url: DATA,
    exchanges: [
      dedupExchange,
      fetchExchange,
      cacheExchange({
        updates: {
          Mutation: {
            MainQuery: (result, args, cache) => {
              cache.invalidate('Query');
            },
          },
        },
      }),
    ],
  }),
});
```

From the "Mutation" keyword i would assume that this invalidation should only happen for mutations, but it seems that queries don't cache now

**Reply by JoviDeCroock** [reply]
Because the order of your exchanges is odd, switch cache and fetch like in the docs

**Reply by Geddard** [reply]
> Because the order of your exchanges is odd, switch cache and fetch like in the docs

Heh the order is not mentioned anywhere, checking

**Reply by Geddard** [reply]
> Because the order of your exchanges is odd, switch cache and fetch like in the docs

with the order changed the cache works, but it's not invalidated

**Reply by kitten** [reply]
As I mentioned before the mutation field will have to be used. `MainQuery` doesn't sound like a field on the `Mutation` type.

**Reply by Geddard** [reply]
> As I mentioned before the mutation field will have to be used. `MainQuery` doesn't sound like a field on the `Mutation` type.

i don't understand.

This is literally the query I'm using, it works perfectly on GraphiQL

```
mutation {
  MainQuery {
    CodeSearch(search: "asd") {
       value
     }
  }
}
```

i tried calling `cache.invalidate` with MainQuery or CodeSearch, same result always no cache removed

**Reply by Geddard** [reply]
And by the way, if i wanted to use the additionalTypenames, would it be like this?
EDIT: I've realized that the second param is variables, so i updated the function to pass an empty object first, but i get the same results , tried with MainQuery or CodeSearch or both and the cache for this call is not dropped


```
from(
      dataClient
        .mutation<T>(query, {}, { additionalTypeNames: ['MainQuery'] })
        .toPromise()
    );
```

To clarify, the queries to which i do the mutation and from where i do the search afterwards are different... maybe that's a problem for urql too?

examples:


```
the queries would be executed in this order

{
  MainQuery {
    CodeSearch(search: "asd") {
      value // returns asd002 as suggestion because asd001 exists
    }
  }
}

mutation {
  MainQuery {
    UpdateSource(input: { code: "asd002" }) {
       output {
         code
       }
     }
  }
}

{
  MainQuery {
    CodeSearch(search: "asd") {
      value // should return asd003 since i just updated the DB with a new code, but i get the cached results from the 1st query
    }
  }
}

```

**Comment by Geddard** [comment, marked as answer]
It took me a while to understand how to fix this, here's the solution
https://github.com/FormidableLabs/urql/discussions/1339#discussioncomment-311501
